substrate/client/network/sync/src/strategy.rs

```diff
@@ -0,0 +1,489 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! [`SyncingStrategy`] is a proxy between [`crate::engine::SyncingEngine`]
+//! and specific syncing algorithms.
+
+pub mod chain_sync;
+mod state;
+pub mod state_sync;
+pub mod warp;
+
+use crate::{
+	types::{BadPeer, OpaqueStateRequest, OpaqueStateResponse, SyncStatus},
+	LOG_TARGET,
+};
+use chain_sync::{ChainSync, ChainSyncAction, ChainSyncMode};
+use libp2p::PeerId;
+use log::{error, info};
+use prometheus_endpoint::Registry;
+use sc_client_api::{BlockBackend, ProofProvider};
+use sc_consensus::{BlockImportError, BlockImportStatus, IncomingBlock};
+use sc_network_common::sync::{
+	message::{BlockAnnounce, BlockData, BlockRequest},
+	SyncMode,
+};
+use sp_blockchain::{Error as ClientError, HeaderBackend, HeaderMetadata};
+use sp_consensus::BlockOrigin;
+use sp_runtime::{
+	traits::{Block as BlockT, NumberFor},
+	Justifications,
+};
+use state::{StateStrategy, StateStrategyAction};
+use std::sync::Arc;
+use warp::{EncodedProof, WarpProofRequest, WarpSync, WarpSyncAction, WarpSyncConfig};
+
+/// Corresponding `ChainSync` mode.
+fn chain_sync_mode(sync_mode: SyncMode) -> ChainSyncMode {
```

Comment:
Is this function really necessary? It reverted https://github.com/paritytech/substrate/pull/14465 and the fact that there are two types again makes it painful to backport https://github.com/paritytech/substrate/pull/14482 (whose goal is to allow combining Substrate's sync with a custom sync mechanism we have in Subspace).

Can this be replaced with a method on `SyncMode` that returns `true` for both `SyncMode::Full` and `SyncMode::Warp`? Or match enum against two possible variants explicitly.

Reply:
What this and upcoming PRs aim to achieve is to clearly separate syncing strategies like `Warp`, `State`, `ChainSync`, etc. Introducing an atomic variable accessible by `ChainSync` that refers to warp sync would be going into the opposite direction.

On the other hand, once the refactoring is over, it should be possible to plug custom sync as a separate strategy and switch to `ChainSync` only when the initial sync is over, without the need to pause `ChainSync`. May be this will suite you?